readme
# =========================================================================== #
# NG CMS // Плагины // Блок онлайн		                                      #
# =========================================================================== #



Плагин добавляет на сайт блок с количеством посетителей, которые зашли на сайте.

Возможности плагина:
- Подсчет количества посетителей, гостей, роботов и суммарное количество посетителей.
- Выбирать тему отображения hint
- Отображение местоположения посетителя на сайте.
- Отображение аватара пользователя.
- Определение IP.
- Определение операционной системы.
- Определение браузера

----------------------------------------
Для вывода плагина добавить в main.tpl переменную {block_online}

----------------------------------------
Шаблон user_block.tpl
Настройка вывода блока посетителей:
----------------------------------------

* sum_count - суммарное количество посетителей
* user_count - количество зарегистрированных пользователей
* guest_count - количество гостей

* online_user - вывод пользователей
* online_bot - вывод роботов

* online_user_list - вывод последних посетителей (за указаное время в настройках)
* online_user_vizit - вывод сегодняшних посетителей (за указаное время в настройках)

----------------------------------------
Шаблон user_hint.tpl 
Настройка вывода hinta пользователя:
----------------------------------------

{hint} - выбор темы hinta
{foto} - аватар пользователя
{ip} - показывает ИП посетителя
{user} - посетитель
{user_OS} - показывает ОС посетителя
{browser_icon} - показывает браузер посетителя
{usergroup} - отображает группу посетителя
{last_visit} - время входа посетителя
{user_location} - где находится на сайте посетитель

----------------------------------------
Файл user-agent.ini
Настройка роботов:
----------------------------------------
Добавить в новых можно настроить в файле user-agent.ini


history
[0.4][2023-04-18]
» добавлен GeoLiteCity (флаг, страна, код)
» добавлены настройки отображения для GeoLiteCity
» добавлены иконки

[0.3][2023-03-21]
» правильно отображение группы пользователя для вновь созданной группы
» добавлены гости сайта
» подправлены местоположения посетителя на сайте

[0.2][2023-03-16]
» Добавлено количество роботов
» Сумарное количиство пользователей и роботов
» Изменения в локации пользователя

[0.1][2023-03-09]
+ Выпуск первой версии

config.php
<?php

# protect against hack attempts
if (!defined('NGCMS')) die ('HAL');

pluginsLoadConfig();
LoadPluginLang('block_online', 'config', '', '', '#');

switch ($_REQUEST['action']) {
	case 'about':			about();		break;
	default: main();
}

function about()
{global $twig, $lang, $breadcrumb;
	$tpath = locatePluginTemplates(array('main', 'about'), 'block_online', 1);
	$breadcrumb = breadcrumb('<i class="fa fa-user-secret btn-position"></i><span class="text-semibold">'.$lang['block_online']['block_online'].'</span>', array('?mod=extras' => '<i class="fa fa-puzzle-piece btn-position"></i>'.$lang['extras'].'', '?mod=extra-config&plugin=block_online' => '<i class="fa fa-user-secret btn-position"></i>'.$lang['block_online']['block_online'].'',  '<i class="fa fa-exclamation-circle btn-position"></i>'.$lang['block_online']['about'].'' ) );

	$xt = $twig->loadTemplate($tpath['about'].'about.tpl');
	$tVars = array();
	$xg = $twig->loadTemplate($tpath['main'].'main.tpl');
	
	$about = 'версия 0.2';
	
	$tVars = array(
		'global' => 'О плагине',
		'header' => $about,
		'entries' => $xt->render($tVars)
	);
	
	print $xg->render($tVars);
}

function main()
{global $twig, $lang, $breadcrumb;
	
	$tpath = locatePluginTemplates(array('main', 'general.from'), 'block_online', 1);
	$breadcrumb = breadcrumb('<i class="fa fa-user-secret btn-position"></i><span class="text-semibold">'.$lang['block_online']['block_online'].'</span>', array('?mod=extras' => '<i class="fa fa-puzzle-piece btn-position"></i>'.$lang['extras'].'', '?mod=extra-config&plugin=block_online' => '<i class="fa fa-user-secret btn-position"></i>'.$lang['block_online']['block_online'].'' ) );

	if (isset($_REQUEST['submit'])){
		pluginSetVariable('block_online', 'limit', intval($_REQUEST['limit']));
		pluginSetVariable('block_online', 'vtimeout', intval($_REQUEST['vtimeout']));
		pluginSetVariable('block_online', 'timeout', intval($_REQUEST['timeout']));
		pluginSetVariable('block_online', 'time_clear', intval($_REQUEST['time_clear']));
		pluginSetVariable('block_online', 'hint', $_REQUEST['hint']);
		pluginSetVariable('block_online', 'icon', $_REQUEST['icon']);
		pluginSetVariable('block_online', 'num_user_last', intval($_REQUEST['num_user_last']));
		pluginSetVariable('block_online', 'num_user_vizit', intval($_REQUEST['num_user_vizit']));
		pluginSetVariable('block_online', 'separator', $_REQUEST['separator']);
		pluginSetVariable('block_online', 'robo_ip', (int)$_REQUEST['robo_ip']);
		pluginSetVariable('block_online', 'robo_geo', (int)$_REQUEST['robo_geo']);
		pluginSetVariable('block_online', 'guest_ip', (int)$_REQUEST['guest_ip']);
		pluginSetVariable('block_online', 'guest_geo', (int)$_REQUEST['guest_geo']);
		
		pluginSetVariable('block_online', 'localsource', (int)$_REQUEST['localsource']);
		pluginsSaveConfig();
		msg(array("type" => "info", "info" => "сохранение прошло успешно"));
		return print_msg( 'info', ''.$lang['block_online']['block_online'].'', 'Cохранение прошло успешно', 'javascript:history.go(-1)' );
	}
	
	$limit = pluginGetVariable('block_online', 'limit');
	$vtimeout = pluginGetVariable('block_online', 'vtimeout');
	$timeout = pluginGetVariable('block_online', 'timeout');
	$time_clear = pluginGetVariable('block_online', 'time_clear');
	$separator = pluginGetVariable('block_online', 'separator');

	$hint = pluginGetVariable('block_online', 'hint');
	$hint = '<option value="hintbox" '.($hint==0?'selected':'').'>по умолчанию</option><option value="hints" '.($hint==1?'selected':'').'>белая</option><option value="hintbox" '.($hint==2?'selected':'').'>темная</option>';
	$icon = pluginGetVariable('block_online', 'icon');
	$icon = '<option value="0" '.($icon==0?'selected':'').'>'.$lang['noa'].'</option><option value="1" '.($icon==1?'selected':'').'>'.$lang['yesa'].'</option>';

	$num_user_last = pluginGetVariable('block_online', 'num_user_last');
	$num_user_last = '<option value="1" '.($num_user_last==1?'selected':'').'>по умолчанию</option><option value="2" '.($num_user_last==1?'selected':'').'>2 столбца</option><option value="3" '.($num_user_last==3?'selected':'').'>3 столбца</option>';

	$num_user_vizit = pluginGetVariable('block_online', 'num_user_vizit');
	$num_user_vizit = '<option value="1" '.($num_user_vizit==1?'selected':'').'>по умолчанию</option><option value="2" '.($num_user_vizit==1?'selected':'').'>2 столбца</option><option value="3" '.($num_user_vizit==3?'selected':'').'>3 столбца</option>';

	$robo_ip = pluginGetVariable('block_online', 'robo_ip');
	$robo_ip = '<option value="0" '.($robo_ip==0?'selected':'').'>'.$lang['noa'].'</option><option value="1" '.($robo_ip==1?'selected':'').'>'.$lang['yesa'].'</option>';
	$robo_geo = pluginGetVariable('block_online', 'robo_geo');
	$robo_geo = '<option value="0" '.($robo_geo==0?'selected':'').'>'.$lang['noa'].'</option><option value="1" '.($robo_geo==1?'selected':'').'>'.$lang['yesa'].'</option>';

	$guest_ip = pluginGetVariable('block_online', 'guest_ip');
	$guest_ip = '<option value="0" '.($guest_ip==0?'selected':'').'>'.$lang['noa'].'</option><option value="1" '.($guest_ip==1?'selected':'').'>'.$lang['yesa'].'</option>';
	$guest_geo = pluginGetVariable('block_online', 'guest_geo');
	$guest_geo = '<option value="0" '.($guest_geo==0?'selected':'').'>'.$lang['noa'].'</option><option value="1" '.($guest_geo==1?'selected':'').'>'.$lang['yesa'].'</option>';

	$xt = $twig->loadTemplate($tpath['general.from'].'general.from.tpl');
	$xg = $twig->loadTemplate($tpath['main'].'main.tpl');
	
	$tVars = array(
		'limit' 		=> $limit,
		'vtimeout' 		=> $vtimeout,
		'timeout' 		=> $timeout,
		'time_clear'	=> $time_clear,
		'icon' 			=> $icon,
		'hint' 			=> $hint,
		'separator' 	=> $separator,
		'robo_ip'   	=> $robo_ip,
		'robo_geo'   	=> $robo_geo,
		'guest_ip'   	=> $guest_ip,
		'guest_geo'   	=> $guest_geo,
		'num_user_last' => $num_user_last,
		'num_user_vizit' => $num_user_vizit,
		'localsource'   => MakeDropDown(array(0 => 'Шаблон сайта', 1 => 'Плагина'), 'localsource', (int)pluginGetVariable('block_online', 'localsource')),
	);
	
	$tVars = array(
		'global' => 'Общие',
		'header' => '<i class="fa fa-exclamation-circle"></i> <a href="?mod=extra-config&plugin=block_online&action=about">'.$lang['block_online']['about'].'</a>',
		'entries' => $xt->render($tVars)
	);
	
	print $xg->render($tVars);
}


install.php
<?php

# protect against hack attempts
if (!defined('NGCMS')) die ('HAL');

function plugin_block_online_install($action) {
	global $lang;
	
	if ($action != 'autoapply')
		loadPluginLang('block_online', 'config', '', '', ':');
		
	$db_create = array(
		array(
			'table' => 'online',
			'action' => 'cmodify',
			'key' => 'primary key (`session`)',
			'fields' => array(
				array('action' => 'cmodify', 'name' => 'session', 'type' => 'varchar(35)', 'params' => 'NOT NULL DEFAULT \'\''),
				array('action' => 'cmodify', 'name' => 'id', 'type' => 'int(11)', 'params' => 'UNSIGNED NOT NULL DEFAULT 0'),
				array('action' => 'cmodify', 'name' => 'lasttime', 'type' => 'int(10)', 'params' => 'UNSIGNED NOT NULL DEFAULT 0'),
				array('action' => 'cmodify', 'name' => 'ip', 'type' => 'varchar(30)', 'params' => 'NOT NULL DEFAULT \'\''),
				array('action' => 'cmodify', 'name' => 'agent', 'type' => 'varchar(255)', 'params' => 'NOT NULL DEFAULT \'\''),
				array('action' => 'cmodify', 'name' => 'os', 'type' => 'varchar(255)', 'params' => 'NOT NULL DEFAULT \'\''),
				array('action' => 'cmodify', 'name' => 'browser', 'type' => 'varchar(255)', 'params' => 'NOT NULL DEFAULT \'\''),
				array('action' => 'cmodify', 'name' => 'login', 'type' => 'varchar(30)', 'params' => 'NOT NULL DEFAULT \'\''),
				array('action' => 'cmodify', 'name' => 'status', 'type' => 'tinyint(1)', 'params' => 'NOT NULL DEFAULT 0'),
				array('action' => 'cmodify', 'name' => 'avatar', 'type' => 'varchar(100)', 'params' => ''),
				array('action' => 'cmodify', 'name' => 'reg', 'type' => 'int(10)', 'params' => 'NOT NULL DEFAULT 0'),
				array('action' => 'cmodify', 'name' => 'location', 'type' => 'varchar(255)', 'params' => 'NOT NULL DEFAULT \'\'')
			)
		)
	);

	switch ($action) {
		case 'confirm': 
			 generate_install_page('block_online', $lang['block_online:install']);
			 break;
		case 'autoapply':
		case 'apply':
			if (fixdb_plugin_install('block_online', $db_create, 'install', ($action=='autoapply')?true:false)) {
				plugin_mark_installed('block_online');
				pluginSetVariable('block_online', 'separator', ',');
				pluginSetVariable('block_online', 'limit', '20');
				pluginSetVariable('block_online', 'vtimeout', '3600');
				pluginSetVariable('block_online', 'timeout', '300');
				pluginSetVariable('block_online', 'time_clear', '3600');
				pluginsSaveConfig();
			} else {
				return false;
			}
			break;
	}
	return true;
}


block_online.php
<?php

# protect against hack attempts
if (!defined('NGCMS')) die ('HAL');

add_act('index', 'block_online');

$BOOT = false;
$jouser = Array();

include_once(dirname(__FILE__) . '/lib/online.class.php');
include_once(dirname(__FILE__) . '/lib/geoipcity.inc');

function isBot()
{
	$uar		= parse_ini_file('lib/user-agent.ini');
	$bot_name	= '';
	
	foreach($uar as $key => $value)
		if(stripos($_SERVER['HTTP_USER_AGENT'], $key)){
			$bot_name = $value;
			break;
		}
	
	if ($bot_name)
		$GLOBALS['BOOT'] = true;
	else
		$bot_name = $_SERVER['HTTP_USER_AGENT'];
	
	return $bot_name;
}

function block_online ()
{
	global $mysql, $tpl, $config, $template, $userROW, $ip, $lang, $UGROUP;
	
	$tpath = locatePluginTemplates(array('skins/user_hint', 'skins/user_block'), 'block_online', intval(pluginGetVariable('block_online', 'localsource')));	

	register_htmlvar('css', admin_url . '/plugins/block_online/tpl/skins/css/online.css');
	register_htmlvar('js', admin_url . '/plugins/block_online/tpl/skins/css/online.js');
	
	LoadPluginLang('block_online', 'main', '', '', ':');
	
	$agent 	= !$userROW['id']    ? isBot()    : $_SERVER['HTTP_USER_AGENT'];
	$login 	= $userROW['name']   ? $userROW['name']   : '';
	$grp 	= $userROW['status'] ? $userROW['status'] : ($GLOBALS['BOOT'] ? -1 : 0);
	$id 	= $userROW['id']     ? $userROW['id']     : 0;
	$os 	= user_OS($_SERVER["HTTP_USER_AGENT"])   ? user_OS($_SERVER["HTTP_USER_AGENT"])   : 'неизвестно';
	$browser 	= user_browser($_SERVER["HTTP_USER_AGENT"])   ? user_browser($_SERVER["HTTP_USER_AGENT"])   : 'неизвестно';
	$location 	= user_position()   ? user_position()   : '';
	$sess = md5($agent.$ip.$id);
	
	# check session
	$ss = $mysql->record('SELECT `session` FROM '.prefix.'_online WHERE `session` = '.db_squote($sess).' LIMIT 1');
	
	$db_column = array('session', 'id', 'lasttime', 'ip', 'agent', 'os', 'browser', 'login', 'status', 'avatar', 'location', 'reg');
	$db_value = array(db_squote($sess), $id, time(), db_squote($ip), db_squote($agent), db_squote($os), db_squote($browser), db_squote($login), $grp, db_squote($userROW['avatar']), db_squote($location), db_squote($userROW['reg']));
	
	if ($ss == false){
		$mysql->query('INSERT INTO `'.prefix.'_online` ('.implode(',', $db_column).') VALUES ('.implode(',', $db_value).')');
	}else{
		$mysql->query('UPDATE `'.prefix.'_online` SET `lasttime`='.time().', `id`='.$id.', `os`='.db_squote($os).', `browser`='.db_squote($browser).', `login`='.db_squote($login).', `status`='.$grp.', `location`='.db_squote($location).' WHERE `session`='.db_squote($ss['session']));
	}

	$i = 0; $u = 0; $b = 0; $g = 0; $list = ''; $obot = ''; $guest = '';
	
	$separator = pluginGetVariable('block_online', 'separator') ? pluginGetVariable('block_online', 'separator') : ', ';

	$num_user_last = pluginGetVariable('block_online', 'num_user_last') ? pluginGetVariable('block_online', 'num_user_last') : 1;
	$num_user_vizit = pluginGetVariable('block_online', 'num_user_vizit') ? pluginGetVariable('block_online', 'num_user_vizit') : 1;
			
	if($num_user_last == 1){
		$separat = $separator;
	}

	if($num_user_vizit == 1){
		$vseparat = $separator;
	}
	
	$timeout = pluginGetVariable('block_online', 'timeout') ? pluginGetVariable('block_online', 'timeout') : 300;
	$tm = time() - intval($timeout);

	$res = $mysql->select('SELECT * FROM `'.prefix.'_online` WHERE `lasttime` > '.$tm.';');
	foreach ($res as $row)
	{
		
		$gi = geoip_open( extras_dir."/block_online/lib/GeoLiteCity.dat", GEOIP_STANDARD );
		$geoip = geoip_record_by_addr( $gi, $row['ip'] );
		geoip_close( $gi );

		if ($row['status'] == -1)
		{
			$b++;
			$k.= '<div class="online_out'.$num_user_last.'">'.$row['agent'].''.$separat.'</div>';
			
			if (is_array($userROW) && ($userROW['status'] == 1 || $userROW['status'] == 2)) { 
				if (pluginGetVariable('block_online', 'robo_ip')){
					$robo_ip = '<b>IP:</b> <span>'.$row['ip'].'</span><br />';
				}
				
				if (pluginGetVariable('block_online', 'robo_geo')){
					$flag = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/flags/".$geoip->country_code."0.gif>";
					$robot4 = '<b>Страна:</b> '.$flag.' '.$geoip->country_name.' '.$geoip->country_code3.'<br />';
				}
			}
			
			$last_date = date('H:i:s', intval($row['lasttime']));
			
			if (pluginGetVariable('block_online', 'icon')){
				$ico_loc = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/read.gif>&nbsp;";
				$ico_vis = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/time.gif>&nbsp;";
			}
			
			$robots = '<div><span style=\'padding:30px 30px 0 30px\'>'.robots($row['agent']).'</span></div>';
			$robot1 = '<b>Браузер:</b> '.$row['browser'].'<br />';
			$robot2 = '<b>Локация:</b> '.$ico_loc.''.$row['location'].'<br />';
			$robot3 = '<b>Последний визит:</b>&nbsp;'.$ico_vis.''.langdate("j Q Y", $last_time).'&nbsp;('.$last_date.')<br />';

			$hint = pluginGetVariable('block_online', 'hint') ? pluginGetVariable('block_online', 'hint') : 'hintbox';

			$bot.= '<a href="" udata="<div class=\''.$hint.'\'><div class=\'lcol\'>'.$robots.'</div><div class=\'rcol\'>'.$robo_ip.''.$robot1.''.$robot2.''.$robot3.''.$robot4.'</div></div>">'.$row['agent'].'</a>'.$separator.'';
		}

		elseif ($row['status'] > 0)
		{
			if ($jouser[$row['id']]) continue;
			
			$status = isset($UGROUP[$row['status']]) ? $UGROUP[$row['status']]['name'] : ('Unknown ['.$row['status'].']');

			$u++;
			
			$profile_link = checkLinkAvailable('uprofile', 'show')?
				 generateLink('uprofile', 'show', array('name' => $row['login'], 'id' => $row['id'])):
				 generateLink('core', 'plugin', array('plugin' => 'uprofile', 'handler' => 'show'), array('id' => $row['id']));
			$avatar_link = $row['avatar'] ? avatars_url.'/'.$row['avatar'] : avatars_url.'/noavatar.gif';
			
			if (is_array($userROW) && ($userROW['status'] == 1 || $userROW['status'] == 2)) { 
				//$userip = $ip ? $ip : $_SERVER["REMOTE_ADDR"];
				$user_ip = '<b>IP:</b>&nbsp;'.$row['ip'].'<br />';
			}
			
			if (is_array($userROW) && ($userROW['status'] == 1 || $userROW['status'] == 2)) { 
				$flag = "<img src=".admin_url."/plugins/block_online/img/flags/".$geoip->country_code."0.gif>";
				$city = $geoip->city;
				$country = $geoip->country_name;
			}
			
			$time = time() + ($config['date_adjust'] * 60);
			$last_time = $time - 500;
			
			$last_date = date('H:i:s', intval($row['lasttime']));
			
			if (pluginGetVariable('block_online', 'icon')){
				$ico_loc = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/read.gif>&nbsp;";
				$ico_vis = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/time.gif>&nbsp;";
			}
			
			$tvars['vars'] = array(
				'hint'			=> pluginGetVariable('block_online', 'hint') ? pluginGetVariable('block_online', 'hint') : 'hintbox',
				'foto'			=> $avatar_link,
				'profile'		=> $profile_link,
				'flag'			=> $flag,
				'city'			=> $city,
				'country'		=> $country,
				'user'			=> $row['login'],
				'user_location'	=> '<b>Локация:</b>&nbsp;'.$ico_loc.''.$row['location'].'<br />',
				'user_agent'	=> $row['os'],
				'ip'			=> $user_ip,
				'browser_icon'	=> '<b>Браузер:</b>&nbsp;'.$row['browser'].'<br />',
				'user_OS'		=> '<b>ОС:</b>&nbsp;'.$row['os'].'<br />',
				'usergroup'		=> '<b>Группа:</b>&nbsp;'.$status.'<br />',
				'last_visit'	=> '<b>Последний визит:</b>&nbsp;'.$ico_vis.''.langdate("j Q Y", $last_time).'&nbsp;('.$last_date.')<br />',
			);
		
			$tpl->template('skins/user_hint', $tpath['skins/user_hint']);
			$tpl->vars('skins/user_hint', $tvars);
			$entries = $tpl->show('skins/user_hint');
			
			$listall = '<div class="online_out'.$num_user_last.'"><a href="'.$profile_link.'" udata="'.$entries.'">'.$row['login'].'</a>'.$separat.'</div>';
			$list .= substr($listall, 0, strlen($listall));
			
			$onuser = '<a href="'.$profile_link.'" udata="'.$entries.'">'.$row['login'].'</a>'.$separator.'';
			$ouser .= substr($onuser, 0, strlen($onuser));

			$jouser[$row['id']] = true;

		} else 
				if ($row['status'] == 0)
		{
			$t.= '<div class="online_out'.$num_user_last.'">гость'.$separat.'</div>';
				$g++;
			
			if (is_array($userROW) && ($userROW['status'] == 1 || $userROW['status'] == 2)) {
				if (pluginGetVariable('block_online', 'guest_ip')){
					$guest_ip = '<div class=\'statonline\'>'.$row['ip'].'</div>';
				}
				if (pluginGetVariable('block_online', 'guest_geo')){
					$flag = "<img src=".admin_url."/plugins/block_online/img/flags/".$geoip->country_code."0.gif>";
					$guest4 = '<b>Страна:</b> '.$flag.' '.$geoip->country_name.' '.$geoip->country_code3.'<br />';
				}
			}
			
			$last_date = date('H:i:s', intval($row['lasttime']));

			if (pluginGetVariable('block_online', 'icon')){
				$ico_loc = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/read.gif>&nbsp;";
				$ico_vis = "<img style=vertical-align:middle; src=".admin_url."/plugins/block_online/img/time.gif>&nbsp;";
			}
			
			$guests = '<img src=\''.admin_url.'/plugins/block_online/img/nouser.png\'>'.$guest_ip.'';
			$guest1 = '<b>Браузер:</b> '.$row['browser'].'<br />';
			$guest2 = '<b>Локация:</b> '.$ico_loc.''.$row['location'].'<br />';
			$guest3 = '<b>Последний визит:</b>&nbsp;'.$ico_vis.''.langdate("j Q Y", $last_time).'&nbsp;('.$last_date.')<br />';

			$hint = pluginGetVariable('block_online', 'hint') ? pluginGetVariable('block_online', 'hint') : 'hintbox';
			$guest.= '<a href="" udata="<div class=\''.$hint.'\'><div class=\'lcol\'>'.$guests.'</div><div class=\'rcol\'>'.$guest1.''.$guest2.''.$guest3.''.$guest4.'</div></div>">гость</a>'.$separator.'';
		}
	}
	
	$lim = pluginGetVariable('block_online', 'limit');
	$limit = $lim ? $lim : 20;
	
	$vtimeout = pluginGetVariable('block_online', 'vtimeout') ? pluginGetVariable('block_online', 'vtimeout') : 3600;
	$vtm = time() - intval($vtimeout);

	$vres = $mysql->select('SELECT * FROM `'.prefix.'_online` WHERE `lasttime` > '.$vtm.' ORDER BY lasttime DESC LIMIT '. $limit.'');
	foreach ($vres as $row)
	{
		if ($row['status'] == -1)
		{
			$vk.= '<div class="online_out'.$num_user_vizit.'">'.$row['agent'].''.$vseparat.'</div>';
		}
		elseif ($row['status'] > 0)
		{
			
			$status = isset($UGROUP[$row['status']]) ? $UGROUP[$row['status']]['name'] : ('Unknown ['.$row['status'].']');

			$vizit = str_replace(array('{profile_link}', '{login}', '{status}', '{num_user_vizit}', '{location}', '{reg}', '{separator}'), array($profile_link, $row['login'], $status, $num_user_vizit, $row['location'], langdate("j Q Y", $row['reg']), $vseparat), $lang['block_online:vizit']);
			
		} else ;

	}
	

	$list = $list.$k;
	$online_all = substr($list, 0, strlen($list)-2);
	$online_user = substr($ouser, 0, strlen($ouser));
	$obot = $bot;
	$online_bot = substr($obot, 0, strlen($obot));
	$oguest = $guest;
	$online_guest = substr($oguest, 0, strlen($oguest));

	$vizit = $vizit.$vk/* .$t */;
	$online_user_vizit = substr($vizit, 0, strlen($vizit)-2);
	
	$tvars['vars'] = array (
		'online_user_list' 	=> $online_all,
		'online_user_vizit' => $online_user_vizit,
		'user_count' 		=> $u,
		'guest_count'		=> $g,
		'bot_count'			=> $b,
		'sum_count'			=> $u + $g + $b,
		'online_user' 		=> $online_user ? $online_user : 'нет',
		'online_bot' 		=> $online_bot ? $online_bot : 'нет',
		'online_guest' 		=> $online_guest ? $online_guest : 'нет',
		'online_time'		=> gmdate('H:i', $vtimeout),
	);

	$tpl->template('skins/user_block', $tpath['skins/user_block']);
	$tpl->vars('skins/user_block', $tvars);
	$template['vars']['block_online'] = $tpl->show('skins/user_block');

	# clear old records
	if (pluginGetVariable('block_online', 'last_clear') < time() - intval(pluginGetVariable('block_online', 'time_clear'))){
		$mysql->query('DELETE FROM `'.prefix.'_online` WHERE `lasttime` < '.$tm.';');
		pluginSetVariable('block_online', 'last_clear', time());
		pluginsSaveConfig();
	}
}

?>